# Basic `pandas`

In this notebook we will learn about `pandas` a vital data management package in python.

By the end of this notebook you will know about:
- `pandas` `Series` and `DataFrame` objects,
- Indexing `pandas` objects,
- Reading files with `pandas`,
- Saving files with `pandas` and 
- Some helpful `pandas` functionality.

## `pandas`

`pandas` is one of the most popular data handling packages in `python`. We will cover the minimum you need to know about the package for the boot camp in this notebook.

Let's start by importing the package. Just like in the previous notebook this should check whether or not you have `pandas` installed on your machine. If you are using Anaconda, <a href="https://www.anaconda.com/">https://www.anaconda.com/</a>, `pandas` should be installed already. If not check out the documentation for installation instructions <a href="https://pandas.pydata.org/docs/getting_started/install.html">https://pandas.pydata.org/docs/getting_started/install.html</a>.

In [1]:
## it is standard to import pandas as pd
import pandas as pd

In [2]:
## let's check what version of numpy you have
## when I wrote this I had version 1.3.5
## yours may be different
print(pd.__version__)

1.3.5


If you had a version of `pandas` installed, both of those code chunks should have executed without error. If not, you will need to install it onto your machine because we will be using it heavily in the boot camp. If you are unsure how to install a python package in general check our python package installation guide, <a href="https://www.erdosinstitute.org/data-science">https://www.erdosinstitute.org/data-science</a>.

##### Be sure you can run both of the above code chunks before continuing with this notebook, again it should be fine if your package version is slightly different than mine.

### `Series` and `DataFrame`s

`pandas` has two main data structures: 
1. `Series` objects, <a href="https://pandas.pydata.org/docs/reference/api/pandas.Series.html">https://pandas.pydata.org/docs/reference/api/pandas.Series.html</a> and 
2. `DataFrame` objects, <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html</a>. 

Let's explore them below.

In [3]:
## We can turn a list into a series
## with pd.Series()
print("Below is a list.")
print([0,1,2,3], type([0,1,2,3]))
print()
print("Below is a pandas Series object.")
print(pd.Series([0,1,2,3]), type(pd.Series([0,1,2,3])))

Below is a list.
[0, 1, 2, 3] <class 'list'>

Below is a pandas Series object.
0    0
1    1
2    2
3    3
dtype: int64 <class 'pandas.core.series.Series'>


The second thing we printed was a `Series` object. Note the two columns of numbers. The first column is the index of the object, the second column contains the values of the object. We can access those two separately like below.

In [4]:
## The index
pd.Series([0,1,2,3]).index

RangeIndex(start=0, stop=4, step=1)

In [5]:
## The values
pd.Series([0,1,2,3]).values

array([0, 1, 2, 3])

In [7]:
## You practice
## Take the list labeled a and 
## turn it into a Series named b
a = [5,2,3,6,'a','b','e',True,False]



b = pd.Series(a)
b

0        5
1        2
2        3
3        6
4        a
5        b
6        e
7     True
8    False
dtype: object

In [8]:
## We can get select entries from a Series 
## using the index
b[3]

6

In [11]:
## Note that we don't have to have our index be a list of numbers
c = pd.Series([1,2,3,4], index=['a','b','c','d'])

print(c)

a    1
b    2
c    3
d    4
dtype: int64


In [13]:
## and we can access by index in the same way
c["d"]

4

Now let's check out a `DataFrame`. `DataFrame`s are essentially a collection of `Series` with a common index. We can also think of them as a table of values with column and row labels.

In [14]:
## We can make a DataFrame using a dictionary
## the dictionary keys are the column labels
## the dictionary values are columns
df = pd.DataFrame({'one':[3,4,5,2,4,5], 
                    'two':['a','b','e','h','l','p'],
                    'third column':[7,7,7,7,7,7]})

## NOTE that this is not the only way to make 
## a dataframe!

print(df)

   one two  third column
0    3   a             7
1    4   b             7
2    5   e             7
3    2   h             7
4    4   l             7
5    5   p             7


This is a `DataFrame`, the unlabeled column is the index, the labeled columns are `Series` objects themselves. We can access them in the following way.

In [15]:
## If your column's name doesn't violate a couple of format rules
## you can use
## df.column_name
print(df.one) 
print()
print("Note that each column is a Series object.")
print(type(df.one))

0    3
1    4
2    5
3    2
4    4
5    5
Name: one, dtype: int64

Note that each column is a Series object.
<class 'pandas.core.series.Series'>


In [16]:
## However if our column name has spaces or certain characters in it
## (like . , ! ? "" and so on)
## we can't use .column_name
df.third column

SyntaxError: invalid syntax (40332708.py, line 4)

In [17]:
## So we have to use df["column name"] instead
# or df['column_name']
print(df['third column']) 
print()
print(type(df['third column']))

0    7
1    7
2    7
3    7
4    7
5    7
Name: third column, dtype: int64

<class 'pandas.core.series.Series'>


In [18]:
## Just like with a Series we can use .index
## to get the index
df.index

RangeIndex(start=0, stop=6, step=1)

In [19]:
## .values returns a 2-D numpy ndarray with our columns
df.values

array([[3, 'a', 7],
       [4, 'b', 7],
       [5, 'e', 7],
       [2, 'h', 7],
       [4, 'l', 7],
       [5, 'p', 7]], dtype=object)

In [20]:
## You code
## Make a data frame, call it my_df 
## Make the first column labeled 'first' from a
## Make the second column labeled 'second' from b
## see what happens when you add , index=range(10,10+len(a)) 
## after the dictionary
a = [4,5,3,4,5,6,0]
b = ['a','c','d','g','l','m','p']


my_df = pd.DataFrame({'first':a,
                         'second':b}, index=range(10,10+len(a)))


In [21]:
my_df

first second
10      4      a
11      5      c
12      3      d
13      4      g
14      5      l
15      6      m
16      0      p

#### Locating `DataFrame` Entries With the Index

Locating entries in a `DataFrame` is slightly more complicated than what we had to do for `Series` objects.

In [22]:
## You code
## try to index the dataframe df like you would
## a list or Series to find the entry corresponding
## to index 1
df[1]

## What happens?

KeyError: 1

In order to get particular `DataFrame` entries you have to use one of two methods, `.loc[]` or `.iloc[]`.

In [23]:
df

one two  third column
0    3   a             7
1    4   b             7
2    5   e             7
3    2   h             7
4    4   l             7
5    5   p             7

In [24]:
## for example we can get the 1 row of df like so with loc
df.loc[1]

one             4
two             b
third column    7
Name: 1, dtype: object

In [25]:
## or like so with iloc
df.iloc[1]

one             4
two             b
third column    7
Name: 1, dtype: object

The difference between `.loc` and `.iloc` is that `.iloc` is restricted to integer based indexing, while `.loc` is more versatile.

This means that if your index is a list of strings you cannot use `.iloc` with just the index's normal value. For example try to run the following.

In [26]:
## Recall what c is
c

a    1
b    2
c    3
d    4
dtype: int64

In [27]:
## now try to get the ath entry using .iloc
c.iloc["a"]

TypeError: Cannot index by location index with a non-integer key

In [28]:
## instead we'd have to notice that the "a" entry is
## the 0 entry if the index were a normal integer based index
c.iloc[0]

1

In [29]:
## However, we can just insert "a" into .loc
c.loc["a"]

1

Another difference is that `.loc` can be used to perform boolean indexing (eg. find me all the rows where the first column is between $2$ and $4$)

In [30]:
df

one two  third column
0    3   a             7
1    4   b             7
2    5   e             7
3    2   h             7
4    4   l             7
5    5   p             7

In [31]:
## give me all the rows where the "one" column is between 2 and 4
df.loc[(df.one < 4) & (df.one > 2)]

one two  third column
0    3   a             7

In [32]:
## we can even go one step further and find a specific column
df.loc[(df.one < 4) & (df.one > 2), 'third column']

0    7
Name: third column, dtype: int64

In [33]:
## or a subset of many columns
## in this case you enter the subset as a list of column name strings
df.loc[(df.one < 4) & (df.one > 2),['two','third column']]

two  third column
0   a             7

In [34]:
## or a subset of many columns
## in this case you enter the subset as a list of column name strings
df.loc[(df.one < 4) & (df.one > 2), ['third column','two']]

third column two
0             7   a

### Reading in a `.csv` file with `pandas`

You can also read in common data file types with `pandas`. Let's load the following.

In [35]:
## You can read in a csv with
## pd.read_csv("filename.csv")
## https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html

## We'll work with the following dataframe in the next section
jr_shots = pd.read_csv("JR_Smith_Shots_2015_16.csv")

In [36]:
jr_shots

LOC_X  LOC_Y  SHOT_MADE_FLAG
0     -106    244               0
1      -96     97               0
2       30     23               0
3     -204     -1               0
4      -76    237               0
..     ...    ...             ...
844    145    -15               0
845   -241     67               0
846    164    195               0
847      0      1               1
848    -89    288               0

[849 rows x 3 columns]

In [37]:
## You code
## read in the file "beer.csv" using pandas
## call it beer_df
beer_df = pd.read_csv("beer.csv")


## you'll practice with this dataframe in the next section

In [38]:
beer_df

Beer_Name  IBU   ABV  Rating Beer_Type
0     Guinness Draught   45   4.2   3.792     Stout
1      Breakfast Stout   60   8.3   4.145     Stout
2     Milk Stout Nitro   25   6.0   3.951     Stout
3        Dragon’s Milk   31  11.0   4.062     Stout
4         Old Rasputin   75   9.0   4.018     Stout
..                 ...  ...   ...     ...       ...
342     Citra Ass Down   68   8.2   3.922       IPA
343  Golden Nugget IPA   45   6.8   4.000       IPA
344           Le Freak  101   9.2   3.778       IPA
345        White Rajah   70   6.8   4.015       IPA
346   Wrought Iron IPA   80   6.9   3.554       IPA

[347 rows x 5 columns]

There are other useful `read` functions like `read_table` and `read_json`, you will see those in the `pandas` practice problems.

### Helpful `pandas` functions

`pandas` offers some really nice built in functions to help you explore any data set you're dealing with. Let's explore them below.

In [41]:
## df.head(n) let's you inspect the first n rows of the dataframe
## n defaults to 5
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html
jr_shots.head()

LOC_X  LOC_Y  SHOT_MADE_FLAG
0   -106    244               0
1    -96     97               0
2     30     23               0
3   -204     -1               0
4    -76    237               0

In [44]:
## You code 
## investigate what .tail(n) does 
## using the beer_df
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.tail.html
## Hint: n must be an integer
beer_df.tail()


Beer_Name  IBU  ABV  Rating Beer_Type
342     Citra Ass Down   68  8.2   3.922       IPA
343  Golden Nugget IPA   45  6.8   4.000       IPA
344           Le Freak  101  9.2   3.778       IPA
345        White Rajah   70  6.8   4.015       IPA
346   Wrought Iron IPA   80  6.9   3.554       IPA

In [62]:
## You code
## investigate what .sample(n) does
## using the beer_df
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sample.html
## Hint: n must be an integer
beer_df.sample(10)


Beer_Name  IBU   ABV  Rating  \
80                               Fade To Black, Vol. 1   30   8.5   3.844   
129      B.O.R.I.S. The Crusher Oatmeal Imperial Stout   60   9.4   4.053   
28                            Peanut Butter Milk Stout   30   5.3   4.003   
203                                     Mosaic Promise   50   5.5   3.783   
151  The Chocolate Manifesto - Triple Chocolate Mil...   60  10.0   4.122   
88                               Quilter's Irish Death   12   7.8   3.905   
99                                         Sea Monster   65  10.0   3.911   
276                                Commodore Perry IPA   70   7.7   3.605   
143                              Big Bear Black Stout™   55   8.1   3.776   
257                                      Red's Rye IPA   70   6.6   3.779   

    Beer_Type  
80      Stout  
129     Stout  
28      Stout  
203       IPA  
151     Stout  
88      Stout  
99      Stout  
276       IPA  
143     Stout  
257       IPA

In [72]:
## You code
## what happens when you add the argument
## random_state=440 in .sample()?
## rerun the code chunk multiple times with or without the random_state
## try different numbers for the random_state

beer_df.sample(random_state=440)


Beer_Name  IBU  ABV  Rating Beer_Type
301  Yellow Rose   62  6.8   4.073       IPA

In [73]:
## df.info() tells you useful information about your
## dataframe, like the column names,
## the number of rows,
## the number of non-empty entries for each column,
## the data type of each column
## and how much memory the dataframe uses
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html
jr_shots.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 849 entries, 0 to 848
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   LOC_X           849 non-null    int64
 1   LOC_Y           849 non-null    int64
 2   SHOT_MADE_FLAG  849 non-null    int64
dtypes: int64(3)
memory usage: 20.0 KB


In [74]:
## df.describe()
## provides the 5 stat breakdown of each numeric column
## that is any column that consists of integers or floats
## will have its, minimum, maximum, first quartile, median,
## third quartile, and mean provided
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html
jr_shots.describe()

LOC_X       LOC_Y  SHOT_MADE_FLAG
count  849.000000  849.000000      849.000000
mean    -6.204947  112.106007        0.414605
std    157.201943   91.254429        0.492944
min   -250.000000  -46.000000        0.000000
25%   -158.000000   21.000000        0.000000
50%     -2.000000  110.000000        0.000000
75%    133.000000  193.000000        1.000000
max    245.000000  693.000000        1.000000

In [77]:
## You code
## Try out .mean() and .max() and .min()
## on beer_df
## Compare the results to .describe()'s output

## .mean() - https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.mean.html
## .max() - https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.max.html
## .min() - https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.min.html

print(".mean()")
print(beer_df.mean())
print()

print(".max()")
print(beer_df.max())
print()

print(".min()")
print(beer_df.min())
print()


print(beer_df.describe())

print()
print(beer_df.info())

.mean()
IBU       57.579251
ABV        7.639769
Rating     3.853980
dtype: float64

.max()
Beer_Name    reDANKulous
IBU                  250
ABV                 18.0
Rating             4.541
Beer_Type          Stout
dtype: object

.min()
Beer_Name    120 Minute IPA
IBU                       6
ABV                     3.8
Rating                3.022
Beer_Type               IPA
dtype: object

              IBU         ABV      Rating
count  347.000000  347.000000  347.000000
mean    57.579251    7.639769    3.853980
std     24.100858    2.216800    0.228915
min      6.000000    3.800000    3.022000
25%     40.000000    6.200000    3.700500
50%     60.000000    7.000000    3.823000
75%     70.000000    8.850000    3.982500
max    250.000000   18.000000    4.541000

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 347 entries, 0 to 346
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Beer_Name  347 non-null    object 
 1  

/var/folders/58/ybrfn44n059260sdf33phdqc0000gn/T/ipykernel_7334/1205141796.py:11: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  print(beer_df.mean())


In [78]:
## What about variables that aren't numeric? Like
## classes
## You can use df.value_counts() for those
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.value_counts.html
jr_shots.SHOT_MADE_FLAG.value_counts()

0    497
1    352
Name: SHOT_MADE_FLAG, dtype: int64

The above tells us that 497 observations have a `SHOT_MADE_FLAG` of 0 and 352 have a `SHOT_MADE_FLAG` of 1.

In [80]:
## You code
## How many of each "Beer_Type" are there in beer_df?
beer_df.Beer_Type.value_counts()

## What happens when you input the argument normalize=True into value_counts()?
beer_df.Beer_Type.value_counts(normalize=True)



IPA      0.56196
Stout    0.43804
Name: Beer_Type, dtype: float64

One task that we will do quite often in the boot camp is split one data set into two non-overlapping data sets. Let's see how we can do that here.

In [81]:
## Let's first randomly sample 100 observations from jr_shots
## Note the .copy()
## this creates a deep copy of the DataFrame 
## instead of a shallow copy (see notebook 5. Shallow and Deep Copies)
jr_1 = jr_shots.sample(100).copy()

In [82]:
## To create jr_2 we'll just drop the indices of jr_1 from
## jr_shots using .drop()
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop.html

## Again we include .copy() to make a deep copy of the dataframe
jr_2 = jr_shots.drop(jr_1.index).copy()

In [83]:
jr_1

LOC_X  LOC_Y  SHOT_MADE_FLAG
105    -14     18               1
490     78     85               0
684   -117    237               1
327   -192    190               0
194    -11    144               1
..     ...    ...             ...
36     222    115               1
845   -241     67               0
282   -127    219               0
319      0      1               1
100   -230     16               1

[100 rows x 3 columns]

In [84]:
jr_2

LOC_X  LOC_Y  SHOT_MADE_FLAG
0     -106    244               0
1      -96     97               0
2       30     23               0
4      -76    237               0
5       25     23               1
..     ...    ...             ...
843     46    252               0
844    145    -15               0
846    164    195               0
847      0      1               1
848    -89    288               0

[749 rows x 3 columns]

In [85]:
## Let's check that it worked like we thought it would
len(jr_1) + len(jr_2) == len(jr_shots)

True

In [86]:
## You code
## split beer_df into beer_1 and beer_2
## beer_1 should be a random sample of 150 observations
## beer_2 should be all rows not in beer_1
beer_1 = beer_df.sample(150).copy()
beer_2 = beer_df.drop(beer_1.index).copy()

In [87]:
len(beer_1) + len(beer_2) == len(beer_df)

True

### Saving a `DataFrame` to File

Just like we easily read in a data file we can use `pandas` to quickly save data to a file.

This is done with the function `df.to_csv()`, <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html</a>.

In [88]:
## Let's first make a data frame
df = pd.DataFrame({'a':[1,2,3],'b':[2,4,6],'c':[17,34,51]})

df

a  b   c
0  1  2  17
1  2  4  34
2  3  6  51

In [89]:
## Now we call df.to_csv(file_name)
df.to_csv("our_first_dataframe.csv")

Go ahead and check your repository, you should now see `our_first_dataframe.csv` in there.

In [90]:
## You code
## read in "our_first_dataframe.csv" using read_csv
## then look at the df

pd.read_csv("our_first_dataframe.csv")


## is anything off?

Unnamed: 0  a  b   c
0           0  1  2  17
1           1  2  4  34
2           2  3  6  51

When you ran `to_csv()` the default is to record the index as well. That is why your `DataFrame` has an unlabeled column when you read it back in using `read_csv`. If we wanted to avoid writing the index to file you include the argument `index=False`. 

In [91]:
## Make the DataFrame again
df = pd.DataFrame({'a':[1,2,3],'b':[2,4,6],'c':[17,34,51]})

df

a  b   c
0  1  2  17
1  2  4  34
2  3  6  51

In [92]:
## write it to file without the index
df.to_csv("our_first_dataframe.csv", index=False)

In [93]:
## You code
## read in "our_first_dataframe.csv" using read_csv
## then look at the df again

pd.read_csv("our_first_dataframe.csv")



a  b   c
0  1  2  17
1  2  4  34
2  3  6  51

That's all you'll need to know from `pandas` to have a firm footing for our boot camp. We will learn more stuff as we go along, but when that time comes you should be ready!

--------------------------

This notebook was written for the Erd&#337;s Institute C&#337;de Data Science Boot Camp by Matthew Osborne, Ph. D., 2022.

Any potential redistributors must seek and receive permission from Matthew Tyler Osborne, Ph.D. prior to redistribution. Redistribution of the material contained in this repository is conditional on acknowledgement of Matthew Tyler Osborne, Ph.D.'s original authorship and sponsorship of the Erdős Institute as subject to the license (see License.md)